### PR TITLE
posix: implement `sysconf`

### LINKED ISSUE #56670: posix: implement `sysconf()`
**Is your enhancement proposal related to a problem? Please describe.**
The [`sysconf()`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sysconf.html) utility is actually quite fantastic to have available for any POSIX environment. For the most part, it is used to have a portable way to query standard system limits that do not change after initialization (including those that are compile-time constant).

**Describe the solution you'd like**
An implementaiton of [`sysconf()`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sysconf.html), possibly with a space-optimized version 

This is one of those interesting cases where you could provide a "small, medium, and large" configuration, cc @stephanosio. 

_XS: O(1) time, O(1) space_
Completed entirely via macros. Not POSIX-ly correct. Would be compile-time const only. Likely would not even compile if an invalid argument is provided, which would be kind of desireable. Would not work for dynamic queries - e.g. `int n; sysconf(n)`.

```cpp
// <unistd.h> or somewhere else more appropriate
// ...
#define __z_posix_sysconf_SC_CLK_TCK CONFIG_SYS_CLOCK_TICKS_PER_SEC
#define __z_posix_sysconf_SC_AIO_MAX (-1) // not supported (could be guarded by some Kconfig)
// ...
#define sysconf(x) __z_posix_sysconf ## x

// application
__ASSERT_NO_MSG(sysconf(_SC_CLK_TCK) == CONFIG_SYS_CLOCK_TICKS_PER_SEC);
__ASSERT_NO_MSG(sysconf(_SC_AIO_MAX) == -1);
__ASSERT_NO_MSG(sysconf(42) == 73); // nonsense - will trigger a compile error
```

Basically free and carefree. This should probably be the default.

_S: O(n) time, O(n) space_
Linear search through the n enabled options, where it's possible that n < k, where k is the minimum number of supported sysconf keys. Not POSIX-ly correct. Not able to distinguish between an invalid value and a value that is not supported.

_M: O(1) time, O(m) space_
LUT made of the m maximum sysconf options that Zephyr supports, where m >= k, the minimum number required to be POSIX-ly correct. This method is able to distinguish between invalid and unsupported arguments.

While this is constant is size, it is spatially unstable if the key-space is sparse. Specifically, that would be the case when the libc did not use a contiguous key space (starting at 0, ideally).

It would probably be desireable in this case to have a `BUILD_ASSERT()` that checks the size of the table (implies the key-space is contiguous).

To give an idea of the space requirements, there are currently 125 required sysconf keys. On 64-bit systems, a long is 8 bytes, so this would be 1000 bytes alone.

_L: O(log(m)) time, O(m) space_
A binary search through the m maximum sysconf options that Zephyr supports. Same definition as above for k. This method is able to distinguish between invalid and unsupported arguments. It supports a sparse set of keys.

The `bsearch()` standard library call can be used here. If keys cannot be sorted at build time in a compile-time const way, then there can be a `SYS_INIT()` call that reorders them in RAM, at the expense of RAM.

**Describe alternatives you've considered**
Not implementing the sysconf POSIX feature.

**Additional context**
https://pubs.opengroup.org/onlinepubs/9699919799/functions/sysconf.html
https://man7.org/linux/man-pages/man3/sysconf.3.html

### PR DESCRIPTION
**Pull request by ycsin**
Add a (pretty much empty) implementation for posix `sysconf`, can be slowly updated as we improve the support for POSIX.

Also added simple API test in ztest & a sample to print everything.

Sample's output:
```
*** Booting Zephyr OS build zephyr-v3.4.0-630-gff2d9cfcb615 ***

Print everything from sysconf:

[errno: 000] _SC_2_C_BIND(0) = -1
[errno: 088] _SC_2_C_DEV(1) = -1
[errno: 088] _SC_2_CHAR_TERM(2) = -1
[errno: 088] _SC_2_FORT_DEV(3) = -1
[errno: 088] _SC_2_FORT_RUN(4) = -1
[errno: 088] _SC_2_LOCALEDEF(5) = -1
[errno: 088] _SC_2_PBS(6) = -1
[errno: 088] _SC_2_PBS_ACCOUNTING(7) = -1
[errno: 088] _SC_2_PBS_CHECKPOINT(8) = -1
[errno: 088] _SC_2_PBS_LOCATE(9) = -1
[errno: 088] _SC_2_PBS_MESSAGE(10) = -1
[errno: 088] _SC_2_PBS_TRACK(11) = -1
[errno: 088] _SC_2_SW_DEV(12) = -1
[errno: 088] _SC_2_UPE(13) = -1
[errno: 088] _SC_2_VERSION(14) = -1
[errno: 088] _SC_ADVISORY_INFO(15) = -1
[errno: 088] _SC_AIO_LISTIO_MAX(16) = -1
[errno: 088] _SC_AIO_MAX(17) = -1
[errno: 088] _SC_AIO_PRIO_DELTA_MAX(18) = -1
[errno: 088] _SC_ARG_MAX(19) = -1
[errno: 088] _SC_ASYNCHRONOUS_IO(20) = -1
[errno: 088] _SC_ATEXIT_MAX(21) = -1
[errno: 088] _SC_BARRIERS(22) = 200809
[errno: 000] _SC_BC_BASE_MAX(23) = -1
[errno: 088] _SC_BC_DIM_MAX(24) = -1
[errno: 088] _SC_BC_SCALE_MAX(25) = -1
[errno: 088] _SC_BC_STRING_MAX(26) = -1
[errno: 088] _SC_CHILD_MAX(27) = -1
[errno: 088] _SC_CLK_TCK(28) = 100
[errno: 088] _SC_CLOCK_SELECTION(29) = -1
[errno: 088] _SC_COLL_WEIGHTS_MAX(30) = -1
[errno: 088] _SC_CPUTIME(31) = -1
[errno: 088] _SC_DELAYTIMER_MAX(32) = -1
[errno: 088] _SC_EXPR_NEST_MAX(33) = -1
[errno: 088] _SC_FSYNC(34) = -1
[errno: 088] _SC_GETGR_R_SIZE_MAX(35) = -1
[errno: 088] _SC_GETPW_R_SIZE_MAX(36) = -1
[errno: 088] _SC_HOST_NAME_MAX(37) = -1
[errno: 088] _SC_IOV_MAX(38) = -1
[errno: 088] _SC_IPV6(39) = -1
[errno: 088] _SC_JOB_CONTROL(40) = -1
[errno: 088] _SC_LINE_MAX(41) = -1
[errno: 088] _SC_LOGIN_NAME_MAX(42) = -1
[errno: 088] _SC_MAPPED_FILES(43) = -1
[errno: 088] _SC_MEMLOCK(44) = -1
[errno: 088] _SC_MEMLOCK_RANGE(45) = -1
[errno: 088] _SC_MEMORY_PROTECTION(46) = -1
[errno: 088] _SC_MESSAGE_PASSING(47) = -1
[errno: 088] _SC_MONOTONIC_CLOCK(48) = -1
[errno: 088] _SC_MQ_OPEN_MAX(49) = -1
[errno: 088] _SC_MQ_PRIO_MAX(50) = -1
[errno: 088] _SC_NGROUPS_MAX(51) = -1
[errno: 088] _SC_OPEN_MAX(52) = -1
[errno: 088] _SC_PAGE_SIZE(53) = -1
[errno: 088] _SC_PAGESIZE(54) = -1
[errno: 088] _SC_PRIORITIZED_IO(55) = -1
[errno: 088] _SC_PRIORITY_SCHEDULING(56) = -1
[errno: 088] _SC_RAW_SOCKETS(57) = -1
[errno: 088] _SC_RE_DUP_MAX(58) = -1
[errno: 088] _SC_READER_WRITER_LOCKS(59) = -1
[errno: 088] _SC_REALTIME_SIGNALS(60) = -1
[errno: 088] _SC_REGEXP(61) = -1
[errno: 088] _SC_RTSIG_MAX(62) = -1
[errno: 088] _SC_SAVED_IDS(63) = -1
[errno: 088] _SC_SEM_NSEMS_MAX(64) = -1
[errno: 088] _SC_SEM_VALUE_MAX(65) = 32767
[errno: 000] _SC_SEMAPHORES(66) = 200809
[errno: 000] _SC_SHARED_MEMORY_OBJECTS(67) = -1
[errno: 088] _SC_SHELL(68) = -1
[errno: 088] _SC_SIGQUEUE_MAX(69) = -1
[errno: 088] _SC_SPAWN(70) = -1
[errno: 088] _SC_SPIN_LOCKS(71) = -1
[errno: 088] _SC_SPORADIC_SERVER(72) = -1
[errno: 088] _SC_SS_REPL_MAX(73) = -1
[errno: 088] _SC_STREAM_MAX(74) = -1
[errno: 088] _SC_SYMLOOP_MAX(75) = -1
[errno: 088] _SC_SYNCHRONIZED_IO(76) = -1
[errno: 088] _SC_THREAD_ATTR_STACKADDR(77) = 200809
[errno: 000] _SC_THREAD_ATTR_STACKSIZE(78) = 200809
[errno: 000] _SC_THREAD_CPUTIME(79) = -1
[errno: 088] _SC_THREAD_DESTRUCTOR_ITERATIONS(80) = -1
[errno: 088] _SC_THREAD_KEYS_MAX(81) = 5
[errno: 000] _SC_THREAD_PRIO_INHERIT(82) = -1
[errno: 088] _SC_THREAD_PRIO_PROTECT(83) = -1
[errno: 088] _SC_THREAD_PRIORITY_SCHEDULING(84) = 200809
[errno: 000] _SC_THREAD_PROCESS_SHARED(85) = -1
[errno: 088] _SC_THREAD_ROBUST_PRIO_INHERIT(86) = -1
[errno: 088] _SC_THREAD_ROBUST_PRIO_PROTECT(87) = -1
[errno: 088] _SC_THREAD_SAFE_FUNCTIONS(88) = -1
[errno: 088] _SC_THREAD_SPORADIC_SERVER(89) = -1
[errno: 088] _SC_THREAD_STACK_MIN(90) = 0
[errno: 000] _SC_THREAD_THREADS_MAX(91) = -1
[errno: 088] _SC_THREADS(92) = -1
[errno: 088] _SC_TIMEOUTS(93) = -1
[errno: 088] _SC_TIMER_MAX(94) = 5
[errno: 088] _SC_TIMERS(95) = -1
[errno: 088] _SC_TRACE(96) = -1
[errno: 088] _SC_TRACE_EVENT_FILTER(97) = -1
[errno: 088] _SC_TRACE_EVENT_NAME_MAX(98) = -1
[errno: 088] _SC_TRACE_INHERIT(99) = -1
[errno: 088] _SC_TRACE_LOG(100) = -1
[errno: 088] _SC_TRACE_NAME_MAX(101) = -1
[errno: 088] _SC_TRACE_SYS_MAX(102) = -1
[errno: 088] _SC_TRACE_USER_EVENT_MAX(103) = -1
[errno: 088] _SC_TTY_NAME_MAX(104) = -1
[errno: 088] _SC_TYPED_MEMORY_OBJECTS(105) = -1
[errno: 088] _SC_TZNAME_MAX(106) = -1
[errno: 088] _SC_V7_ILP32_OFF32(107) = -1
[errno: 088] _SC_V7_ILP32_OFFBIG(108) = -1
[errno: 088] _SC_V7_LP64_OFF64(109) = -1
[errno: 088] _SC_V7_LPBIG_OFFBIG(110) = -1
[errno: 088] _SC_V6_ILP32_OFF32(111) = -1
[errno: 088] _SC_V6_ILP32_OFFBIG(112) = -1
[errno: 088] _SC_V6_LP64_OFF64(113) = -1
[errno: 088] _SC_V6_LPBIG_OFFBIG(114) = -1
[errno: 088] _SC_VERSION(115) = 200809
[errno: 000] _SC_XOPEN_CRYPT(116) = -1
[errno: 088] _SC_XOPEN_ENH_I18N(117) = -1
[errno: 088] _SC_XOPEN_REALTIME(118) = -1
[errno: 088] _SC_XOPEN_REALTIME_THREADS(119) = -1
[errno: 088] _SC_XOPEN_SHM(120) = -1
[errno: 088] _SC_XOPEN_STREAMS(121) = -1
[errno: 088] _SC_XOPEN_UNIX(122) = -1
[errno: 088] _SC_XOPEN_UUCP(123) = -1
[errno: 088] _SC_XOPEN_VERSION(124) = -1
[errno: 088] unknown(125) = -1
```

reference: https://github.com/apache/nuttx/blob/master/libs/libc/unistd/lib_sysconf.c

fixes #56670

### CODE REVIEW

File: lib/posix/sysconf.c [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+	case _SC_BC_DIM_MAX:
+		break;
+	case _SC_BC_SCALE_MAX:
+		break;
+	case _SC_BC_STRING_MAX:
+		break;
+	case _SC_CHILD_MAX:
+		break;
+	case _SC_CLK_TCK:
+		retval = CONFIG_SYS_CLOCK_TICKS_PER_SEC;
```
Comment (pdgendt) [inline lib/posix/sysconf.c:80]: Missing `errcode = 0;` and `break;`?

Comment (ycsin) [reply lib/posix/sysconf.c:80]: oops, you are right!

File: lib/posix/sysconf.c [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+		break;
+	case _SC_V6_ILP32_OFFBIG:
+		break;
+	case _SC_V6_LP64_OFF64:
+		break;
+	case _SC_V6_LPBIG_OFFBIG:
+		break;
+	case _SC_VERSION:
+		errcode = 0;
+		retval = 200809L;
```
Comment (ycsin) [inline lib/posix/sysconf.c:287]: This should be `SUPPORTED_POSIX1_VERSION` as well

**Comment by cfriedt** [comment]
This should probably be tied very closely to Kconfig and official POSIX feature test macros i.e. the item but without `_SC`.

Additionally, if implemented via macrobatics, it can be done with 0 runtime overhead and zero additional bytes of space (since it mostly becomes compile-time const data).

That approach was loosely described in the O(1) time O(1) space part of the issue.

**Review by cfriedt — CHANGES_REQUESTED** [review]
Would be nice if it were macro based / compile-time const. That should be the default implementation in Zephyr

**Pull request closed without merging**